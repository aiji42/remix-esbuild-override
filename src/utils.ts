export const resolve = (mod: string): string | null => {
  try {
    return require.resolve(mod);
  } catch (_) {
    return null;
  }
};

export const load = <T>(mod: string): T | null => {
  try {
    return require(mod) as T;
  } catch (_) {
    return null;
  }
};
