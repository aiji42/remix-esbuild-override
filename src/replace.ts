import { spawn } from "node:child_process";
import { renameSync, symlinkSync, mkdirSync } from "node:fs";
import { resolvePath } from "./utils";

const esbuildOverrideMainPath = "remix-esbuild-override/dist/index.js";

export const replaceEsbuild = (restartable = true) => {
  const isOverridden = resolvePath("esbuild").endsWith(esbuildOverrideMainPath);
  if (isOverridden) {
    console.log(
      "ðŸ’½ esbuild has already been replaced. Your custom config can be used to build for Remix"
    );
  } else {
    const [originalEsbuildPath] =
      resolvePath("esbuild").match(/^.*\/esbuild/) ?? [];
    const [esbuildOverridePath] =
      resolvePath("remix-esbuild-override").match(
        /^.*\/remix-esbuild-override/
      ) ?? [];

    const replacedOriginalEsbuildPath = originalEsbuildPath.replace(
      /esbuild$/,
      "esbuild-org"
    );

    renameSync(originalEsbuildPath, replacedOriginalEsbuildPath);
    symlinkSync(esbuildOverridePath, originalEsbuildPath);
    mkdirSync(`${esbuildOverridePath}/bin`);
    symlinkSync(
      `${replacedOriginalEsbuildPath}/bin/esbuild`,
      `${esbuildOverridePath}/bin/esbuild`
    );

    console.log(
      "ðŸ’½ Replaced esbuild. Your custom config can be used to build for Remix"
    );

    if (restartable) restartOnce();
  }
};

const restartOnce = () => {
  if (process.env.restarting)
    throw new Error("An unexpected error has occurred.");

  console.log("ðŸ’½ Auto restarting precess...");
  spawn(process.argv[0], process.argv.slice(1), {
    env: { restarting: "true" },
    stdio: "ignore",
  }).unref();
};
