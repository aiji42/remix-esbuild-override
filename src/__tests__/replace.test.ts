import { fs, vol } from "memfs";
import * as utils from "../utils";
import { replaceEsbuild } from "../replace";

vi.mock("../utils");
vi.mock("fs", () => fs);
const mockedConsoleLog = vi.spyOn(console, "log");

beforeEach(() => {
  vi.spyOn(utils, "restartOnce").mockImplementation(vi.fn());
  vol.reset();
});
afterEach(() => {
  vi.resetAllMocks();
});

test("replace esbuild by remix-esbuild-override", () => {
  vi.spyOn(utils, "resolvePath").mockImplementation((moduleName) => {
    if (moduleName === "esbuild")
      return "/app/node_modules/esbuild/lib/main.js";
    if (moduleName === "remix-esbuild-override")
      return "/app/node_modules/remix-esbuild-override/dist/index.js";
    return "";
  });
  vol.fromJSON(
    {
      "./node_modules/esbuild/package.json":
        '{ "name": "esbuild", "main": "lib/main.js" }',
      "./node_modules/esbuild/lib/main.js": `this is original main`,
      "./node_modules/esbuild/bin/esbuild": `this is original binary`,
      "./node_modules/remix-esbuild-override/package.json": `{ "name": "remix-esbuild-override", "main": "dist/index.js" }`,
      "./node_modules/remix-esbuild-override/dist/index.js": `this is replaced main`,
    },
    "/app"
  );

  replaceEsbuild();

  expect(
    vol.readFileSync("/app/node_modules/esbuild/dist/index.js", "utf8")
  ).toEqual("this is replaced main");
  expect(
    vol.readFileSync("/app/node_modules/esbuild/bin/esbuild", "utf8")
  ).toEqual("this is original binary");
  expect(vol.toJSON()).toMatchSnapshot();

  expect(mockedConsoleLog).toBeCalledWith(
    expect.stringMatching(/Replaced esbuild/)
  );
});

test("already replaced", () => {
  vi.spyOn(utils, "resolvePath").mockReturnValue(
    "node_modules/remix-esbuild-override/dist/index.js"
  );

  replaceEsbuild();
  expect(mockedConsoleLog).toBeCalledWith(
    expect.stringMatching(/esbuild has already been replaced/)
  );
});
