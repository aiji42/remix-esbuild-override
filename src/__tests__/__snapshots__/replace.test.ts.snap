// Vitest Snapshot v1

exports[`replace esbuild by remix-esbuild-override 1`] = `
{
  "/app/node_modules/esbuild-org/bin/esbuild": "this is original binary",
  "/app/node_modules/esbuild-org/lib/main.js": "this is original main",
  "/app/node_modules/esbuild-org/package.json": "{ \\"name\\": \\"esbuild\\", \\"main\\": \\"lib/main.js\\" }",
  "/app/node_modules/remix-esbuild-override/dist/index.js": "this is replaced main",
  "/app/node_modules/remix-esbuild-override/package.json": "{ \\"name\\": \\"remix-esbuild-override\\", \\"main\\": \\"dist/index.js\\" }",
}
`;
